### PR TITLE
Add ErrorOptions to RequestOptions

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR adds a new enum `ErrorOptions` to the `com.azure.core.http.rest` package. `ErrorOptions` replaces a boolean that determined whether requests using `RequestOptions` throw when an error response (400 or above) is received. This change allows for finer controlled configurations in the request and allows for new handling to be added in the future.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RequestOptions.java [hunk 1]
```diff
@@ -125,13 +128,15 @@
...
-     * (400 or above) is received.
+     * Gets the {@link ErrorOptions} that determines how error responses (400 or above) are handled.
+     * <p>
+     * Default is to throw.
      *
-     * @return true if to throw on status codes of 400 or above, false if not. Default is true.
+     * @return The {@link ErrorOptions} that determines how error responses (400 or above) are handled. Default is
+     * to throw.
      */
-    boolean isThrowOnError() {
```
Comment (lmolkova) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RequestOptions.java:133]: I assume it was never released, so we can just remove it?

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ErrorOptions.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+package com.azure.core.http.rest;
+
+/**
+ * Determines how errors are handled by requests using {@link RequestOptions}.
+ */
+public enum ErrorOptions {
+    /**
+     * Throw exceptions when an HTTP response with a status code indicating an error (400 or above) is received.
+     */
+    THROW,
```
Comment (joshfree) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ErrorOptions.java:13]: would we ever need to handle 500 or above differently than 400-499?

**Review by joshfree — APPROVED** [review]

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java [hunk 3]
```diff
@@ -417,10 +417,23 @@ private static Exception instantiateUnexpectedException(final UnexpectedExceptio
...
         final int responseStatusCode = decodedResponse.getSourceResponse().getStatusCode();
 
+        boolean requestOptionsNoThrow = false;
+        boolean requestOptionsThrow = false;
+        if (options != null) {
+            requestOptionsNoThrow = options.getErrorOptions().contains(ErrorOptions.NO_THROW);
+            requestOptionsThrow = options.getErrorOptions().contains(ErrorOptions.THROW);
+        }
+
+        if (requestOptionsNoThrow && requestOptionsThrow) {
```
Comment (lmolkova) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:427]: can we check validity in the `RequestOptions.setErrorOptions` instead? I would simplify  getting options (i.e. we can only get `options.getErrorOptions().contains(ErrorOptions.NO_THROW)` and never check THROW

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RequestOptions.java (hunk 1 above)
Comment (JonathanGiles) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RequestOptions.java:133]: Correct

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java (hunk 3 above)
Comment (JonathanGiles) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:427]: Agreed

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RequestOptions.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RequestOptions.java:133]: Correct, this has never been GA'd

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ErrorOptions.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/ErrorOptions.java:13]: That's a great question, the short answer would be maybe. Though, this aligns with default handling that already occurs elsewhere within the SDKs.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:427]: That's a good question, my concern with `setErrorOptions` validating is as we add more options there could be more complex cases where having both could be permissible.

@JonathanGiles any thoughts?

Comment (JonathanGiles) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/RestProxy.java:427]: For now we can probably centralise the validation in the setter, and spread it out into appropriate places if complex cases arise.

**Pull request merged**